text/0000-constants-in-patterns.md

```diff
@@ -0,0 +1,316 @@
+- Nothing else has structural equality.
+
+In particular, the value of `C` must be known at pattern-building time (which is pre-monomorphization).
+
+After ensuring all conditions are met, the constant value is translated into a pattern, and now behaves exactly as-if that pattern had been written directly.
+In particular, it fully participates in exhaustiveness checking.
+(For raw pointers, constants are the only way to write such patterns. Only `_` is ever considered exhaustive for these types.)
+
+Practically speaking, to determine whether a `struct`/`enum` type's `PartialEq` behaves exactly like the one generated by `derive(PartialEq)`, we use a trait:
+
+```rust
+/// When implemented on a `struct` or `enum`, this trait indicates that
+/// the `PartialEq` impl of `Self` behaves exactly like the one
+/// generated by `derive(PartialEq)`:
+/// - on a `struct`, it returns `true` if and only if comparing all
+///   fields with their respective `PartialEq` returns `true`.
+/// - in an `enum`, it returns `true` if and only if both values
+///   have the same variant, and furthermore comparing the fields of
+///   that variant with their respective `PartialEq` returns `true`
+///   for all fields.
+///
+/// This trait should not be implemented on `union` types.
+///
+/// This is a "shallow" property in the sense that it says nothing about
+/// the behavior of `PartialEq` on the fields of this type, it only relates
+/// `PartialEq` on this type to that of its fields.
+///
+/// This trait is used when determining whether a constant may be used as a pattern:
+/// all types appearing in the value of the constant must implement this trait.
+///
+/// All that said, this is a safe trait, so violating these requirements
+/// can only lead to logic bugs or accidentally exposing an equality test that your
+/// library would otherwise not provide, not to unsoundness.
+trait StructuralPartialEq: PartialEq {}
+```
+
+This trait is automatically implemented when writing `derive(PartialEq)`.
+For this RFC to be implemented, the trait can remain unstable can hence cannot be implemented directly by users.
+In the future, it might be possible for libraries that implement `PartialEq` by hand (for instance for performance reasons) to also implement `StructuralPartialEq` by hand, if they can promise that the comparison behaves as documented.
+(See "Future possibilities" for some of the open questions around that option.)
+The trait has `PartialEq` as a supertrait because its entire contract only makes sense for types that implement `PartialEq`.
+
+Range patterns are only allowed on integers, `char`, and floats; for floats, neither end must be a `NaN`.
+
+The *behavior* of such a constant as a pattern is the same as the corresponding native pattern.
+On floats are raw pointers, pattern matching behaves like `==`,
+which means in particular that the value `-0.0` matches the pattern `0.0`, and NaN values match no pattern.
```

Comment:
Is a wildcard considered a pattern?

```suggestion
which means in particular that the value `-0.0` matches the pattern `0.0`, and NaN values only match wildcard patterns (`_`).
```

Reply:
Good catch, we should mention wildcards.